src/scanner.cc

```diff
@@ -34,13 +34,14 @@ struct Scanner {
       Tag &tag = tags[j];
       if (tag.type == CUSTOM) {
         unsigned name_length = tag.custom_tag_name.size();
-        if (i + 2 + name_length >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) break;
+        if (name_length > UINT8_MAX) return 0;
+        if (i + 2 + name_length >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) return 0;
         buffer[i++] = static_cast<char>(tag.type);
         buffer[i++] = name_length;
         tag.custom_tag_name.copy(&buffer[i], name_length);
         i += name_length;
       } else {
-        if (i + 1 >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) break;
+        if (i + 1 >= TREE_SITTER_SERIALIZATION_BUFFER_SIZE) return 0;
```

Comment:
I think this change might be the one causing test failures. I'd like to still preserve as much of the state as possible in the case of HTML files with crazy-deep nesting. Maybe we can just move the serialization of the length to the end of the method, after the loop? Something like:

```cpp
unsigned i = 0;
unsigned tag_count = 0;
i += sizeof(tag_count);
for (unsigned n = tags.size(); tag_count < n; tag_count++) {
  Tag &tag = tags[tag_count];
  // ...
}

// Write the final tag count to the beginning of the buffer
std::memcpy(buffer, &tag_count, sizeof(tag_count));
return i;
```